### PR TITLE
Recursive denormalize with property info

### PR DESCRIPTION
**Pull request by mihai-stancu**
[2.7][Serializer] Feature using `PropertyInfo` to recursively denormalize values before calling the setter.

| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | none |
| License | MIT |
| Doc PR | n/a |

Refactored #14844 "Denormalize with typehinting":
- Now using (optional) PropertyInfo to extract type information
- Updated tests
- Updated composer.json

This PR replaces #16143.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Status: needs work

**Comment by dunglas** [comment]
@mihai-stancu all changes required for this PR are now merged in master. You can now safely rebase and finish it. The code should be moved in the new `AbstractObjectNormalizer` class.

File: src/Symfony/Component/Serializer/Annotation/Groups.php [hunk 1]
```diff
@@ -35,7 +35,7 @@ class Groups
      */
     public function __construct(array $data)
     {
-        if (!isset($data['value']) || !$data['value']) {
+        if (empty($data['value'])) {
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Annotation/Groups.php]: Should be reverted, we use the explicit form in Symfony.

**Comment by dunglas** [comment]
@mihai-stancu can you finish this one? I can do it (I'll keep your commits to give you full credit) if you're busy.

**Comment by mihai-stancu** [comment]
Hi Kevin,

I'm a bit busy indeed, if you could do it this week that would be fine.
Otherwise I'll do it next week.

**Comment by mihai-stancu** [comment]
This work is included in #17660 .

**Pull request closed without merging**